### PR TITLE
APIv2: Revenue metrics

### PR DESCRIPTION
**Pull request by macobo**
### Changes

This PR brings revenue metrics to APIv2. They are still not yet publicly available but are now fully functional for dashboard use.

Notes:
- No behavioral changes for users
- Revenue currencies are stored on `Query` similar to goals with associated business logic
- Revenue metric access is validated properly
- Old behavior of removing revenue metrics when unavailable is done via a Query flag and QueryOptimizer behavior.

Needed to bring breakdowns frontend to APIv2.

### CODE REVIEW

File: lib/plausible/stats/filters/query_parser.ex [hunk 1]
```diff
@@ -405,14 +410,22 @@ defmodule Plausible.Stats.Filters.QueryParser do
     end
   end
 
-  def preload_goals_if_needed(site, filters, dimensions) do
+  def preload_goals_if_needed(site, metrics, filters, dimensions) do
```
Comment (RobertJoonas) [inline lib/plausible/stats/filters/query_parser.ex:413]: A companion to my other suggestion - also rename this function to `preload_needed_goals`.

**Review by RobertJoonas — COMMENTED** [review]
I think I found a small bug here (revenue metrics should show up in the Top Stats as €0.0 instead of blank):

https://pr-4659.review.plausible.io/plausible.io?filters=((is,goal,(Payment)))&period=day&keybindHint=E&date=2024-10-08

Otherwise the PR looks good to me. Left a couple of suggestions inline though.

File: lib/plausible/goals/filters.ex [hunk 2]
```diff
@@ -34,18 +34,37 @@ defmodule Plausible.Goals.Filters do
     end)
   end
 
+  def preload(goals, filters) do
```
Comment (apata) [inline lib/plausible/goals/filters.ex:37]: **suggestion, non-blocking**: When used as Goals.Filters.preload, it's not obvious what records are getting preloaded. Bit more verbose, but how's `Goals.Filters.reduce_goals_by_filters`?

File: extra/lib/plausible/stats/goal/revenue.ex [hunk 3]
```diff
@@ -2,83 +2,74 @@ defmodule Plausible.Stats.Goal.Revenue do
...
+  def available?(site) do
+    site = Plausible.Repo.preload(site, :owner)
+    Plausible.Billing.Feature.RevenueGoals.check_availability(site.owner) == :ok
   end
 
-  def cast_revenue_metrics_to_money(results, _), do: results
+  # :NOTE: Legacy queries don't have metrics associated with them so work around the issue by assuming
+  #   revenue metric was requested.
+  def requested?([]), do: true
+  def requested?(metrics), do: Enum.any?(metrics, &(&1 in @revenue_metrics))
```
Comment (apata) [inline extra/lib/plausible/stats/goal/revenue.ex:66]: **suggestion, non-blocking**: I understand this will be temp, but could we detect the legacy case here by having the query.v2 as a second arg? `Revenue.requested?(query.metrics, query.v2)`

**Review by apata — APPROVED** [review]
Nice work!

File: lib/plausible/goals/filters.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (macobo) [reply lib/plausible/goals/filters.ex]: Good suggestion!

File: lib/plausible/stats/filters/query_parser.ex [hunk 4]
```diff
@@ -460,6 +473,25 @@ defmodule Plausible.Stats.Filters.QueryParser do
     end
   end
 
+  on_ee do
+    alias Plausible.Stats.Goal.Revenue
+
+    defdelegate preload_revenue_currencies(site, preloaded_goals, metrics, dimensions),
+      to: Plausible.Stats.Goal.Revenue
```
[the remark this answers is not in the document]
Comment (macobo) [reply lib/plausible/stats/filters/query_parser.ex:477]: Yes, but I preferred the full name for explicitness. `defdelegate` is quite rare in our codebase so I wanted the code-reader to have a straight-forward experience.

File: lib/plausible/goals/filters.ex (hunk 2 above)
Comment (macobo) [reply lib/plausible/goals/filters.ex]: Robert above had a similar thought. Updated!

File: extra/lib/plausible/stats/goal/revenue.ex (hunk 3 above)
Comment (macobo) [reply extra/lib/plausible/stats/goal/revenue.ex:64]: Doesn't seem straight-forward to introduce it without polluting preload call with extra args as well.

**Comment by macobo** [comment]
@RobertJoonas good catch, this also caught an untested edge case. Fixed and updated the tests.

File: test/plausible_web/controllers/api/stats_controller/top_stats_test.exs [hunk 5]
```diff
@@ -1488,6 +1489,71 @@ defmodule PlausibleWeb.Api.StatsController.TopStatsTest do
...
+        build(:event,
+          name: "Payment",
+          revenue_reporting_amount: Decimal.new(1_000),
+          revenue_reporting_currency: "USD"
+        ),
+        build(:event, name: "Signup"),
+        build(:event, name: "Signup")
+      ])
+
+      filters = Jason.encode!(%{goal: "Payment|Signup"})
```
[the remark this answers is not in the document]
Comment (macobo) [reply test/plausible_web/controllers/api/stats_controller/top_stats_test.exs:1513]: Fair. I used this for the consistency with the rest of the file, but should likely refactor all of it at this point as it doesn't reflect actual dashboard calls. Adding it to my mental TODO.

File: extra/lib/plausible/stats/goal/revenue.ex (hunk 3 above)
Comment (apata) [reply extra/lib/plausible/stats/goal/revenue.ex:64]: Ah, roger that. The idea was to get rid of the note, but unless we passed the query itself as the only arg, it wouldn't be obvious.

**Review by RobertJoonas — APPROVED** [review]
